native/membrane_videocompositor/src/elixir_bridge/scene_validation.rs

```diff
@@ -0,0 +1,411 @@
+use std::collections::{HashMap, HashSet};
+
+use super::elixir_structs::*;
+
+pub struct SceneValidator<'a> {
+    objects: &'a [(ObjectName, Object)],
+    final_object_name: &'a ObjectName,
+}
+
+impl<'a> SceneValidator<'a> {
+    pub fn new(scene: &'a Scene) -> Self {
+        Self {
+            objects: &scene.objects,
+            final_object_name: &scene.output,
+        }
+    }
+
+    fn objects_map(&self) -> HashMap<&ObjectName, &Object> {
+        self.objects
+            .iter()
+            .map(|(name, object)| (name, object))
+            .collect::<HashMap<_, _>>()
+    }
+
+    fn check_for_duplicate_pad_refs(&self) -> Result<(), SceneParsingError> {
+        let mut input_pads = HashSet::new();
+
+        for (_, object) in self.objects {
+            if let Object::Video(InputVideo { input_pad }) = object {
+                if !input_pads.insert(input_pad) {
+                    return Err(SceneParsingError::DuplicatePadReferences(input_pad.clone()));
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn check_for_duplicate_or_undefined_names(&self) -> Result<(), SceneParsingError> {
+        if self.objects.is_empty() {
+            return Err(SceneParsingError::UndefinedName(
+                self.final_object_name.clone(),
+            ));
+        }
+
+        let mut names = HashSet::new();
+        for (name, _) in self.objects {
+            if !names.insert(name) {
+                return Err(SceneParsingError::DuplicateNames(name.clone()));
+            }
+        }
+
+        for (_, object) in self.objects {
+            for name in object.mentioned_names() {
+                if !names.contains(name) {
+                    return Err(SceneParsingError::UndefinedName(name.clone()));
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn check_for_unused_objects(&self) -> Result<(), SceneParsingError> {
+        let defined_names = self
+            .objects
+            .iter()
+            .map(|(name, _)| name)
+            .collect::<HashSet<_>>();
+
+        let mut used_names = self
+            .objects
+            .iter()
+            .flat_map(|(_, object)| object.mentioned_names())
+            .collect::<HashSet<_>>();
+
+        if used_names.contains(self.final_object_name) {
+            return Err(SceneParsingError::CycleDetected);
+        } else {
+            used_names.insert(self.final_object_name);
+        }
+
+        let unused_names = defined_names.difference(&used_names).collect::<Vec<_>>();
+
+        if !unused_names.is_empty() {
+            return Err(SceneParsingError::UnusedObject((*unused_names[0]).clone()));
+        }
+
+        Ok(())
+    }
+
+    /// returns `Err(SceneParsingError::CycleDetected)` if a cycle exists in the scene graph, `Ok(())` otherwise.
```

Comment:
[NIT] This is quite obvious from the function name spec. We should either remove this doc or add docs like this to all those functions for consistency.